packages/astro/src/core/app/node.ts

```diff
@@ -116,11 +117,34 @@ export class NodeApp extends App {
 		}
 		return super.match(req, opts);
 	}
-	render(req: NodeIncomingMessage | Request, routeData?: RouteData, locals?: object) {
+	render(request: NodeIncomingMessage | Request, options?: RenderOptions): Promise<Response>
+	/**
+	 * @deprecated Instead of passing `RouteData` and locals individually, pass an object with `routeData` and `locals` properties.
+	 * See https://github.com/withastro/astro/pull/9199 for more information.
+	 */
+	render(request: NodeIncomingMessage | Request, routeData?: RouteData, locals?: object): Promise<Response>
+	render(req: NodeIncomingMessage | Request, routeDataOrOptions?: RouteData | RenderOptions, locals?: object) {
+		let routeData: RouteData | undefined;
+		
+		if (routeDataOrOptions && ('routeData' in routeDataOrOptions || 'locals' in routeDataOrOptions)) {
+			if ('routeData' in routeDataOrOptions) {
+				routeData = routeDataOrOptions.routeData;
+			}
+			if ('locals' in routeDataOrOptions) {
+				locals = routeDataOrOptions.locals;
+			}
+		}
+		else {
+			routeData = routeDataOrOptions as RouteData | undefined;
+			if (routeDataOrOptions || locals) {
+				super.logRenderOptionsDeprecationWarning();
+			}
+		}
+		
 		if (!(req instanceof Request)) {
 			req = createRequestFromNodeRequest(req);
 		}
-		return super.render(req, routeData, locals);
+		return super.render(req, { routeData, locals });
```

Comment:
I think we can pass this as `super.render(req, routeData, locals);` like before so that the warning handling is all handled within `App`, and we don't have to duplicate the validation in `NodeApp`?

We can also make `logRenderOptionsDeprecationWarning` private with a leading `#` then.

Reply:
Wouldn't that make it log the warning everytime?

Reply:
I'm not sure I understand. Wouldn't the duplicate validation here also log the warning everytime? As long as the consumer of `NodeApp` (in this case `@astrojs/node`) uses the new API form, it shouldn't log the warning.

Reply:
I think my initial suggestion was incorrect. I meant `super.render(req, routeDataOrOptions, locals);`. Perhaps that's where the confusion is 😅 

If we simply pass the arguments back to `super.render`, the newer API would also be passed through.

Reply:
ok yeah that makes sense